### PR TITLE
build: consume internal dependencies by dynamic path in devshells

### PR DESCRIPTION
**Pull request by ysndr**
## Background

Out codebase has grown multiple subsystems that are depended on by flox or depend on other subsystems themselves.
E.g. `flox -> pkgdb -> flox-activation-scripts -> flox-activations` for one of the worse ones. On the one hand it is important to be able to build final artefacts of each individual subsystem with nix to allow for better build caching if nothing else. Nix also makes defining these dependencies relatively easy through e.g. string contexts.

However, since our devshell is derived from _the inputs of the toplevel binaries_, the naive approach results in
1. all dependencies being built upon entering the devshell
2. the devshell being in accessible while any dependency fails to build
3. the devshell being stale as soon as any dependency changes -- requiring us to reenter the devshell

Until not too long ago, most of the now standalone subsystems were consolidated in pkgdb (i.e. buildenv and associated functionality).
In order to avoid the 3 pitfalls above, for development, we would set `PKGDB_BIN` to `pkgdb`, instead of `${flox-pkgdb}/bin/pkgdb`, in turn removing the dependency on pkgdb form the build closure, as well as the dev shell closure.
`Pkgdb` now needed to be built separately from the `flox` binary, which we ensured would happen through via the central Justfile. As we added internal (nix built) dependencies to `pkgdb`, we amended the pkgdb Makefile to build these dependencies as requisites of the `pkgdb` binary and provide their path to the C++ preprocessor, when they were unspecified in dev mode.

While `flox` only depended on binaries this worked out relatively well, except occasionally we would still find internal dependencies in the closure of the devshell. That was because we

1. forgot to add "dev mode" overrides,
2. depend on a package by both `flox` and `pkgdb`,
3. required absolute paths which broke our heuristic of adding binaries to PATH,
4. etc.

As we are phasing out pkgdb as a binary, and move more functionality into independent packages orchestrated by the `flox` binary, we are losing some of the dynamic setup we had implemented via the `pkgdb` Makefile.
For example, most recently we have experienced issues wrt `flox-activation-scripts`, which made its way into the devshell (alongside, flox-buildenv and transitively `pkgdb` ...).

This commit is trying to consolidate the dependency management for `flox` and the remains of pkgdb, by applying a similar strategy at an earlier stage.

## Proposed Changes

1. packages **only** provide build environment variables that reference dependencies iff the dependency is provided, i.e. not overridden to `null`
2. within the `floxDevelopmentPackages` scope, all internal dependencies of internal packages are overridden to `null.
3. define all environment variables that point to internal dependencies within the `shellHook` of the devshell and point them to a local directory. `cargo` built dependencies become: `FLOX_ACTIVATIONS_BIN="''${REPO_ROOT}/cli/target/debug/flox-activations"` `nix` built dependencies become `FLOX_BUILDENV="''${REPO_ROOT}/build/flox-buildenv"`
4. define Justfile targets that build dependencies with either `cargo` or `nix` thereby populating the directories previously defined in the environment.

Compared to the previous Makefile implementation this avoids requiring to rebuild binaries as the dependencies change, because everything will depend on stable paths that are updated in place.

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -114,8 +114,49 @@
           flox-cli-tests = callPackage ./pkgs/flox-cli-tests { };
         };
 
+      overlays.development = final: prev: {
+        floxDevelopmentPackages = prev.lib.makeScope prev.newScope (self: {
```
Comment (zmitchell) [inline flake.nix:118]: Question: can you explain the usage of scopes? From searching around it looks like it creates a new collection of packages to draw from in addition to nixpkgs in a way that the packages in the scope can use each other as inputs?

File: pkgs/flox-activation-scripts/default.nix [hunk 2]
```diff
@@ -38,7 +38,11 @@ let
     # Note that substitution doesn't work with variables containing "-"
     # so we need to create and use alternative names.
     process_compose = process-compose;
-    flox_activations = flox-activations;
+    flox_activations =
+      if flox-activations != null then
+        "${flox-activations}/bin/flox-activations"
+      else
+        "$FLOX_ACTIVATIONS_BIN";
```
Comment (mkenigs) [inline pkgs/flox-activation-scripts/default.nix:45]: blocking: can you run the resulting artifact outside of the devShell where `$FLOX_ACTIVATIONS_BIN` isn't set?

File: pkgs/flox-cli-tests/default.nix [hunk 3]
```diff
@@ -177,13 +177,18 @@ writeShellScriptBin PROJECT_NAME ''
...
       "export FLOX_ACTIVATIONS_BIN='${FLOX_ACTIVATIONS_BIN}';"
   }
   ${if FLOX_BIN == null then "export FLOX_BIN='flox';" else "export FLOX_BIN='${FLOX_BIN}';"}
   export PROCESS_COMPOSE_BIN='${process-compose}/bin/process-compose';
-  export FLOX_INTERPRETER='${flox-activation-scripts}';
+  ${
+    if flox-activation-scripts == null then
+      ''export FLOX_INTERPRETER="$PROJECT_ROOT_DIR/build/flox-activation-scripts";''
+    else
+      ''export FLOX_INTERPRETER='${flox-activation-scripts}';''
```
Comment (mkenigs) [inline pkgs/flox-cli-tests/default.nix:190]: nonblocking suggestions:
- do we need any of these exports if they aren't used in any of our bats files? Can we drop all of them except maybe `FLOX_BIN` and `NIX_BIN`?
- Can we get rid of the double defaulting, where there's both `flox-activations` and `FLOX_ACTIVATIONS_BIN ? "${flox-activations}/bin/flox-activations"` as arguments to this file?
- Is there a reason we need all of the `--flox-activations` arguments to this script rather than just falling back to an env var?

**Review by mkenigs — CHANGES_REQUESTED** [review]
One blocking question but otherwise looks good to me

File: Justfile
Comment (dcarley) [inline Justfile:1]: (non-blocking) Do we still need the `.ACTIVATION_SCRIPTS_PACKAGE_DIR` dance in `pkgdb`?

**Review by dcarley — APPROVED** [review]
Confirmed that this solves the `flox-core` issue I had in https://github.com/flox/flox/issues/2371

`just build-cli` has gone from 0.3s to 1.2s when there are no changes, but I'd prefer that small slowdown over all the time we've lost to changes not being picked up and adding individual workarounds.

File: Justfile
Comment (ysndr) [reply Justfile:1]: Most likely not, but I'm not sure how much time we want to invest fixing up pkgdb specifically (vs just removing the buildenv command entirely)

File: pkgs/flox-activation-scripts/default.nix (hunk 2 above)
Comment (ysndr) [reply pkgs/flox-activation-scripts/default.nix:45]: If it's unset the activation scripts won't work no. The rust binaries are different since they pick the env var up at _buildtime_ and have the result compiled in so they don't rely on it at runtime.
The equivalent for the bash scripts would be referring to their dependencies by "relative path to themselves" because nix won't be able to know where they will be put and thus can't substitute an absolute path here.

Is running the results _outside of the shell_ (even with `nix develop -c ...`) blocking?

File: pkgs/flox-cli-tests/default.nix (hunk 3 above)
Comment (ysndr) [reply pkgs/flox-cli-tests/default.nix:190]: There is a lot of duplication in the tests script setup
I believe most of the defaulting can be removed but I tried to keep the changes here to a minimum for this PR to keep the size and scope in check.
I'm also unsure about the impact radius of this. Also tbh, I'd prefer to just run `bats ./cli/tests` instead of going through this wrapper.

File: flake.nix [hunk 4]
```diff
@@ -114,8 +114,49 @@
...
+          rust-internal-deps = prev.rust-internal-deps.override {
+            flox-buildenv = null;
```
[the remark this answers is not in the document]
Comment (ysndr) [reply flake.nix:120]: The main difference is that `flox-cli` depends on `rust-internal-deps` at "eval time".
Setting `rust-internal-deps = null` would affect multiple places where we inherit inputs from `rust-internal-deps` ([1..4]). those allow to propagate all required (external) dependencies in the shell with `mkShell { inputsFrom = [ flox-cli ] }`.


[1] https://github.com/flox/flox/blob/6c63321b2a484ae663de6f13f195efb81672258f/pkgs/flox-cli/default.nix#L119
[2] https://github.com/flox/flox/blob/6c63321b2a484ae663de6f13f195efb81672258f/pkgs/flox-cli/default.nix#L109
[3] https://github.com/flox/flox/blob/6c63321b2a484ae663de6f13f195efb81672258f/pkgs/flox-cli/default.nix#L106
[4] https://github.com/flox/flox/blob/6c63321b2a484ae663de6f13f195efb81672258f/pkgs/flox-cli/default.nix#L96

File: flake.nix (hunk 1 above)
Comment (ysndr) [reply flake.nix:118]: scopes are the nixpkgs intended way to create package groups.

For our case, it mainly provides `callPackage` which you can use to evaluate pacakge expressions with the packages of the scope available (shadowing the same packages in the parent).
We use this below [1] to create the devshell with only the development versions of the packages in scope. the alternative would have been define these as a "normal" attribute set and pulling the pacakges out one by one in the devshell expression.

```nix
# shells/default.nix
{
    flox-cli,
    flox-buildenv,
    flox-package-builder,
    [...]
}: mkShell { [...] }
```

compared to 

```nix
# shells/default.nix
{
   floxDevelopmentPackages
    [...]
}: let 
inherit (floxDevelopmentPackages) 
    flox-cli 
    flox-buildenv
    flox-package-builder
    [...];
in mkShell { [...] } 
```

The latter is still possible with or without scopes if we preferred that.

File: Justfile
Comment (ysndr) [reply Justfile:1]: that make command running with every change because annoying while testing.
Should be fixed now.

File: pkgs/flox-activation-scripts/default.nix (hunk 2 above)
Comment (zmitchell) [reply pkgs/flox-activation-scripts/default.nix:45]: I regularly use the devshell to build something and then test it outside of the devshell because certain things don't work inside the devshell.

File: Justfile
Comment (dcarley) [reply Justfile:1]: Thanks!

File: pkgs/flox-activation-scripts/default.nix (hunk 2 above)
Comment (mkenigs) [reply pkgs/flox-activation-scripts/default.nix:45]: I think people outside the team do so as well

Comment (ysndr) [reply pkgs/flox-activation-scripts/default.nix:45]: well i think the canonical way would be to create a full nix build, using development builds outside of the development environment seems strange.

[6855dc1](https://github.com/flox/flox/pull/2481/commits/6855dc1c28d66ecb73fb17d21c09f3728278e73f) adds some workaround for this situation though.

tl;dr; `./build/wrapper <cmd> <args>` should successfully run `<cmd> <args>` within an environment with the critical values set, e.g. `./build/wrapper flox install hello`

**Review by limeytexan — APPROVED** [review]

File: pkgs/flox-activation-scripts/default.nix (hunk 2 above)
Comment (mkenigs) [reply pkgs/flox-activation-scripts/default.nix:45]: Can we use `"${BASH_SOURCE[0]}/../whatever"` here and get rid of the build wrapper?

**Review by mkenigs — APPROVED** [review]
I would still prefer using BASH_SOURCE in the nix built artifacts rather than with a wrapper so `flox` can be used directly outside of the dev shell but that doesn't need to block

File: pkgs/flox-activation-scripts/default.nix (hunk 2 above)
Comment (ysndr) [reply pkgs/flox-activation-scripts/default.nix:45]: Not without adding multiple variables I think, because the flox activations in particle is called from scripts at different depths. Will check in Monday.

Comment (ysndr) [reply pkgs/flox-activation-scripts/default.nix:45]: `FLOX_ACTIVATIONS_BIN` is currently read by both `activation-scrips/activate` and `activation_scripts/activate.d/{start,attach-inplace}.bash`.
This would mean that we'd need to set it to `${BASH_SOURCE[0]}/../flox-activations/bin/flox-activations` and `${BASH_SOURCE[0]}/../../flox-activations/bin/flox-activations` respectively.

So apart from a wrapper we can, 1) not use `FLOX_ACTIVATIONS_BIN` in one of them, or at least do a better job to pass it to `activate.d/{start,attach-inplace}.bash` unless they need to run independently as well
2) use separate variables, 3) make the "find upwards" logic more complex e.g. `$(while [ $PWD != / ] && [ ! -e $PWD/flake.nix ]; do cd ..; done; if [ $PWD = "/" ]; exit 1; else; echo $PWD; fi)`.

Besides if all we want is a `flox` binary that can be called from anywhere we can also make the wrapper more specific to always exec `FLOX_BIN`, and in doubt using a nix build is still an option -- likely the most reliable one in either case.

Since that's detail id prefer to go forward with this for now and discuss this usecase separately.
